ci_visibility.go

```diff
@@ -0,0 +1,225 @@
+		DefaultValue: "",
+		IsRequired:   false,
+	},
+	"node_hostname": {
+		EnvNames:     []string{"PLUGIN_SYSTEM_HOST", "DRONE_SYSTEM_HOSTNAME", "DRONE_SYSTEM_HOST"},
+		DefaultValue: "",
+		IsRequired:   false,
+	},
+	"node_name": {
+		EnvNames:     []string{"PLUGIN_SYSTEM_HOST", "DRONE_SYSTEM_HOSTNAME", "DRONE_SYSTEM_HOST"},
+		DefaultValue: "",
+		IsRequired:   false,
+	},
+	"node_workspace": {
+		EnvNames:     []string{"PLUGIN_WORKSPACE", "DRONE_WORKSPACE", "HARNESS_WORKSPACE"},
+		DefaultValue: "",
+		IsRequired:   false,
+	},
+	"pipeline_url": {
+		EnvNames:     []string{"PLUGIN_BUILD_LINK", "DRONE_BUILD_LINK", "CI_BUILD_LINK"},
+		DefaultValue: "",
+		IsRequired:   true,
+	},
+	"pipeline_unique_id": {
+		EnvNames:     []string{"PLUGIN_PIPELINE_ID", "HARNESS_PIPELINE_ID"},
+		DefaultValue: "",
+		IsRequired:   true,
+	},
+	"pipeline_status": {
+		EnvNames:     []string{"PLUGIN_BUILD_STATUS", "DRONE_BUILD_STATUS", "CI_BUILD_STATUS"},
+		DefaultValue: "",
+		IsRequired:   true,
+	},
+	"pipeline_type": {
+		EnvNames:     []string{""},
+		DefaultValue: "cipipeline_resource_request",
+		IsRequired:   false,
+	},
+}
+
+func getPipelineVisibilityEnvValue(env string) string {
+	for _, env := range pipelineVisibilityEnvConfigs[env].EnvNames {
+		if value := os.Getenv(env); value != "" {
+			return value
+		}
+	}
+	return pipelineVisibilityEnvConfigs[env].DefaultValue
+}
+
+func ConvertUnixToRFC3339(timestamp int64) string {
+	if timestamp == 0 {
+		return "" // Return an empty string if the timestamp is invalid
+	}
+	t := time.Unix(timestamp, 0).UTC()
+	return t.Format(time.RFC3339)
+}
+
+func parseUnix(value string) int64 {
+	ts, err := strconv.ParseInt(value, 10, 64)
+	if err != nil {
+		return 0 // Return 0 on error
+	}
+	return ts
+}
+
+func ValidateRequiredEnvVars() error {
+	var missing []string
+
+	for key, conf := range pipelineVisibilityEnvConfigs {
+		if conf.IsRequired {
+			val := getPipelineVisibilityEnvValue(key)
+			if val == "" {
+				missing = append(missing, conf.EnvNames...)
+			}
+		}
+	}
+
+	if len(missing) > 0 {
+		return fmt.Errorf("missing required env vars: %s", strings.Join(missing, ", "))
+	}
+	return nil
+}
+
+func SendDatadogPipelineEvent(cfg *Config) error {
+	requestBody := map[string]interface{}{
+		"data": map[string]interface{}{
+			"attributes": map[string]interface{}{
+				"resource": map[string]interface{}{
+					"level":         getPipelineVisibilityEnvValue("pipeline_level"),
+					"unique_id":     getPipelineVisibilityEnvValue("pipeline_unique_id"),
+					"name":          getPipelineVisibilityEnvValue("pipeline_unique_id"),
+					"url":           getPipelineVisibilityEnvValue("pipeline_url"),
+					"start":         ConvertUnixToRFC3339(parseUnix(getPipelineVisibilityEnvValue("pipeline_start"))),
+					"end":           ConvertUnixToRFC3339(parseUnix(getPipelineVisibilityEnvValue("pipeline_end"))),
+					"status":        strings.ToLower(getPipelineVisibilityEnvValue("pipeline_status")),
```

Comment:
no validation against Datadog's accepted values.
- Datadog expects: "success", "error", or "canceled"
- Harness provides: "SUCCESS", "FAILURE", "ABORTED" (need to ensure these map correctly)
```go
func mapStatus(status string) string {
    status = strings.ToLower(status)
    switch status {
    case "success":
        return "success"
    case "failure", "error":
        return "error"
    case "aborted", "canceled":
        return "canceled"
    default:
        return status
    }
}
```